File: algorithms/two-pointers/3sum.js
var threeSum = function(nums) {  
    const results = [];
    const target = 0;
    nums.sort((a,b) => a - b);
  
    for(let i=0; i<nums.length-2; i++){
      if(i > 0 && nums[i] === nums[i-1]) continue;
      let left = i+1;
      let right = nums.length-1;
  
      while(left<right){
        let sum = nums[i]+nums[left]+nums[right];
        
        if(sum > target){ 
          right--;
        }
        else if(sum < target){ 
          left++;
        }
        else{ 
          results.push([nums[i],nums[left],nums[right]])
          while(nums[left] === nums[left+1]) left++;
          while(nums[right] === nums[right-1]) right--;
          left++
          right--
        }
      }
  
    }
    return results;
  };

File: algorithms/two-pointers/3sum.py
from typing import List
# Given an integer array nums
# return all the triplets [nums[i], nums[j], nums[k]] 
# such that i != j, i != k, and j != k
# and nums[i] + nums[j] + nums[k] == 0

# must NOT contain duplicate triplets.
nums = [-1,0,1,2,-1,-4]
#                   h 
#      [-4, -1, -1, 0, 1, 2]  
#            i  l   
target = 0
# Output: [[-1,-1,2],[-1,0,1]]

def threeSum(nums: List[int]) -> List[List[int]]:
    nums.sort()
    result = []
    for i, num in enumerate(nums):
        if nums[i] > 0:
            break
        if nums[i-1] != nums[i]:
            lo = i+1
            high = len(nums)-1
            while lo < high:
                sum = nums[i] + nums[lo] + nums[high]
                if sum > target:
                    high -=1
                elif sum < target:
                    lo+=1
                else:
                    result.append([nums[i], nums[lo], nums[high]])
                    while nums[lo] == nums[lo+1]: lo+=1
                    while nums[high] == nums[high-1]: high-=1
                    lo+=1
                    high-=1

    return result


print(threeSum(nums))